### PR TITLE
Mtc optimize feedupdater

### PR DESCRIPTION
**Pull request by binh-dam-ibigroup**
### Checklist

- [x] Appropriate branch selected => This is a patch on the MTC deployment branch. A similar PR will eventually be issued to the `dev` branch.
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [na] The description lists all applicable issues this PR seeks to resolve
- [na] The description lists any configuration setting(s) that differ from the default settings
- [ ] All tests and CI builds passing => Semantic release does not pick up the correct Node version.

### Description

This PR improves performance and memory usage of the `FeedUpdater` feature:
- the query that populates/updates `versionsToMarkAsProcessed` uses less memory by only requesting the `id` field (and using `FeedVersionSummaries` instead of `FeedVersion`) (see comparison charts below). A bug in the query is also fixed.
- A single mongo query is performed, instead of one per agency, to get a map of all feed sources by agency (what MTC calls "feed id").
- A single mongo query is performed, instead of one per agency, to get a map of the latest feed version sent to publishing for each feed source.

Screenshots of memory profiles (left: current, right: with PR changes):

image

### CODE REVIEW

File: src/test/java/com/conveyal/datatools/manager/jobs/FeedUpdaterTest.java [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+            prop1,
+            prop2,
+            prop3
+        );
+
+        Map<String, String> feedIdsToFeedSourceIds = FeedUpdater.mapFeedIdsToFeedSourceIds(properties);
+
+        assertEquals(FEED_SOURCE_1_ID, feedIdsToFeedSourceIds.get(FIRST_FEED));
+        assertEquals("other-feed-source-id", feedIdsToFeedSourceIds.get(OTHER_FEED));
+        assertNull(feedIdsToFeedSourceIds.get("unknownFeed"));
```
Comment (br648) [inline src/test/java/com/conveyal/datatools/manager/jobs/FeedUpdaterTest.java]: What is the purpose of this test? The outcome will always be null because `feedIdsToFeedSourceIds` will never contain "unknownFeed".

File: src/main/java/com/conveyal/datatools/manager/jobs/FeedUpdater.java [hunk 2]
```diff
@@ -101,12 +111,54 @@ public static FeedUpdater createForTest(CompletedFeedRetriever completedFeedRetr
...
+
+    public static Map<String, FeedSource> mapFeedIdsToFeedSources(
+        Map<String, String> feedIdsToFeedSourceIds,
+        List<FeedSource> feedSources
+    ) {
+        Map<String, FeedSource> feedSourceIdToFeedSource = feedSources.stream()
+            .collect(Collectors.toMap(src -> src.id, Function.identity()));
+
+        return feedIdsToFeedSourceIds.entrySet().stream()
+            .collect(Collectors.toMap(e -> e.getKey(), e -> feedSourceIdToFeedSource.get(e.getValue())));
```
Comment (br648) [inline src/main/java/com/conveyal/datatools/manager/jobs/FeedUpdater.java]: SonarLint: `e -> e.getKey()` can be replaced with `Map.Entry::getKey`. Although it seems clearer to me to leave as is!

File: src/test/java/com/conveyal/datatools/manager/jobs/FeedUpdaterTest.java [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+        List<String> expectedIds = Lists.newArrayList(
+            FIRST_FEED,
+            OTHER_FEED
+        );
+        assertEquals(expectedIds, FeedUpdater.getFeedIds(s3objects));
+    }
+
+    @Test
+    void shouldMapFeedIdsToFeedSourceIds() {
+        ExternalFeedSourceProperty prop1 = new ExternalFeedSourceProperty();
```
Comment (br648) [inline src/test/java/com/conveyal/datatools/manager/jobs/FeedUpdaterTest.java]: Consider creating a helper method to create `ExternalFeedSourceProperty` instances.

File: src/test/java/com/conveyal/datatools/manager/jobs/FeedUpdaterTest.java [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+
+        Map<String, FeedSource> feedIdToFeedSource = FeedUpdater.mapFeedIdsToFeedSources(feedIdsToFeedSourceIds, feedSources);
+        assertEquals(fs1, feedIdToFeedSource.get(FIRST_FEED));
+        assertEquals(fs2, feedIdToFeedSource.get(OTHER_FEED));
+    }
+
+    @Test
+    void shouldGetLatestVersionsSentToPublisher() {
+        // Create two feed sources, and two feed versions in each that have different sentToExternalPublisher values.
+        FeedSource source1 = new FeedSource(FEED_SOURCE_1_ID);
```
Comment (br648) [inline src/test/java/com/conveyal/datatools/manager/jobs/FeedUpdaterTest.java]: Consider creating helper methods to create `FeedSource` and `FeedVersion` instances.

File: src/main/java/com/conveyal/datatools/manager/jobs/FeedUpdater.java [hunk 5]
```diff
@@ -264,18 +322,75 @@ private void updatePublishedFeedVersion(String feedId, FeedVersion publishedVers
...
+            "$replaceRoot": {
+                "newRoot": "$items"
+            }
+        }
+        ])
+        */
+
+        List<String> feedSourceIds = feedSources.stream().map(fs -> fs.id).collect(Collectors.toList());
+
+        List<Bson> stages = Lists.newArrayList(
```
Comment (br648) [inline src/main/java/com/conveyal/datatools/manager/jobs/FeedUpdater.java:359]: Is there potential for this to fail silently if there is a change to the underlying fields? Could this be guarded against perhaps by checking for the existence of all required fields first?

**Review by br648 — COMMENTED** [review]
Just a few comments for consideration. Interesting low level Mongo queries.

File: src/test/java/com/conveyal/datatools/manager/jobs/FeedUpdaterTest.java (hunk 1 above)
Comment (binh-dam-ibigroup) [reply src/test/java/com/conveyal/datatools/manager/jobs/FeedUpdaterTest.java]: That assertion would have covered the uses of the `resourceType` field, but it doesn't seem to be used for anything except the resources tabs jn the feed settings. I removed that assertion.

File: src/main/java/com/conveyal/datatools/manager/jobs/FeedUpdater.java (hunk 2 above)
Comment (binh-dam-ibigroup) [reply src/main/java/com/conveyal/datatools/manager/jobs/FeedUpdater.java]: Updated in https://github.com/ibi-group/datatools-server/pull/508/commits/3bb3b32530a87f525317d0b46cc21e71ed11067a.

File: src/test/java/com/conveyal/datatools/manager/jobs/FeedUpdaterTest.java (hunk 3 above)
Comment (binh-dam-ibigroup) [reply src/test/java/com/conveyal/datatools/manager/jobs/FeedUpdaterTest.java]: Yes, I was on the fence but I've refactored in https://github.com/ibi-group/datatools-server/pull/508/commits/3bb3b32530a87f525317d0b46cc21e71ed11067a.

File: src/test/java/com/conveyal/datatools/manager/jobs/FeedUpdaterTest.java (hunk 4 above)
Comment (binh-dam-ibigroup) [reply src/test/java/com/conveyal/datatools/manager/jobs/FeedUpdaterTest.java]: Refactored in https://github.com/ibi-group/datatools-server/pull/508/commits/3bb3b32530a87f525317d0b46cc21e71ed11067a.

File: src/main/java/com/conveyal/datatools/manager/jobs/FeedUpdater.java (hunk 5 above)
Comment (binh-dam-ibigroup) [reply src/main/java/com/conveyal/datatools/manager/jobs/FeedUpdater.java:359]: The check for the only required field `sentToExternalPublisher` besides `_id` is already being done in the `match` clause.

Comment (br648) [reply src/main/java/com/conveyal/datatools/manager/jobs/FeedUpdater.java:359]: If I change `feedSourceId` or `latestSentToExternalPublisher` to "TEST", Mongo does not throw an SQL exception. Perhaps the test `shouldGetLatestVersionsSentToPublisher` failing is enough to cover this if fields are changed in the future? I understand that this is an edge case, but an RDMS would throw an exception for missing/unknown field.

**Review by br648 — COMMENTED** [review]
Just need your thoughts on this: https://github.com/ibi-group/datatools-server/pull/508#discussion_r1098502741

File: src/main/java/com/conveyal/datatools/manager/jobs/FeedUpdater.java (hunk 5 above)
Comment (binh-dam-ibigroup) [reply src/main/java/com/conveyal/datatools/manager/jobs/FeedUpdater.java:359]: Correct, this has to do with how Mongo fundamentally works. If the field is renamed for some reason, the publishing feature will silently fail, but also the test `shouldGetLatestVersionsSentToPublisher` that ties the Java field with the Mongo field will fail too as you mention, so we should be covered.

**Review by br648 — APPROVED** [review]
All good. We should look to use Mongo queries (as used here) more often.

**Review by philip-cline — APPROVED** [review]
LGTM
